File: go/strassen.go
package main

import "sync"

func strassenSequential(A, B, C [][]int) {
	n := len(A)

	if n <= THRESHOLD {
		multiplyMatrixIterative(A, B, C)
		return
	}

	half := n / 2

	a11, a12, a21, a22 := splitView(A)
	b11, b12, b21, b22 := splitView(B)
	c11, c12, c21, c22 := splitView(C)

	// S matrices (half x half)
	s1 := makeZeroMatrix(half)
	s2 := makeZeroMatrix(half)
	s3 := makeZeroMatrix(half)
	s4 := makeZeroMatrix(half)
	s5 := makeZeroMatrix(half)
	s6 := makeZeroMatrix(half)
	s7 := makeZeroMatrix(half)
	s8 := makeZeroMatrix(half)
	s9 := makeZeroMatrix(half)
	s10 := makeZeroMatrix(half)

	// P matrices
	p1 := makeZeroMatrix(half)
	p2 := makeZeroMatrix(half)
	p3 := makeZeroMatrix(half)
	p4 := makeZeroMatrix(half)
	p5 := makeZeroMatrix(half)
	p6 := makeZeroMatrix(half)
	p7 := makeZeroMatrix(half)

	// Compute S matrices
	sub(b12, b22, s1)
	add(a11, a12, s2)
	add(a21, a22, s3)
	sub(b21, b11, s4)
	add(a11, a22, s5)
	add(b11, b22, s6)
	sub(a12, a22, s7)
	add(b21, b22, s8)
	sub(a11, a21, s9)
	add(b11, b12, s10)

	// Compute P matrices
	strassenSequential(a11, s1, p1)
	strassenSequential(s2, b22, p2)
	strassenSequential(s3, b11, p3)
	strassenSequential(a22, s4, p4)
	strassenSequential(s5, s6, p5)
	strassenSequential(s7, s8, p6)
	strassenSequential(s9, s10, p7)

	// Combine into C
	// C11 = P5 + P4 - P2 + P6
	add(p5, p4, c11)
	sub(c11, p2, c11)
	add(c11, p6, c11)

	// C12 = P1 + P2
	add(p1, p2, c12)

	// C21 = P3 + P4
	add(p3, p4, c21)

	// C22 = P5 + P1 - P3 - P7
	add(p5, p1, c22)
	sub(c22, p3, c22)
	sub(c22, p7, c22)
}

func strassenParallel(A, B, C [][]int) {
	n := len(A)

	if n <= THRESHOLD {
		multiplyMatrixIterative(A, B, C)
		return
	}

	half := n / 2

	a11, a12, a21, a22 := splitView(A)
	b11, b12, b21, b22 := splitView(B)
	c11, c12, c21, c22 := splitView(C)

	// S matrices
	s1 := makeZeroMatrix(half)
	s2 := makeZeroMatrix(half)
	s3 := makeZeroMatrix(half)
	s4 := makeZeroMatrix(half)
	s5 := makeZeroMatrix(half)
	s6 := makeZeroMatrix(half)
	s7 := makeZeroMatrix(half)
	s8 := makeZeroMatrix(half)
	s9 := makeZeroMatrix(half)
	s10 := makeZeroMatrix(half)

	// P matrices
	p1 := makeZeroMatrix(half)
	p2 := makeZeroMatrix(half)
	p3 := makeZeroMatrix(half)
	p4 := makeZeroMatrix(half)
	p5 := makeZeroMatrix(half)
	p6 := makeZeroMatrix(half)
	p7 := makeZeroMatrix(half)

	// Compute S
	sub(b12, b22, s1)
	add(a11, a12, s2)
	add(a21, a22, s3)
	sub(b21, b11, s4)
	add(a11, a22, s5)
	add(b11, b22, s6)
	sub(a12, a22, s7)
	add(b21, b22, s8)
	sub(a11, a21, s9)
	add(b11, b12, s10)

	// Parallelize only if large enough
	if n >= PARALLEL_THRESHOLD {
		var wg sync.WaitGroup
		wg.Add(7)

		go func() { strassenParallel(a11, s1, p1); wg.Done() }()
		go func() { strassenParallel(s2, b22, p2); wg.Done() }()
		go func() { strassenParallel(s3, b11, p3); wg.Done() }()
		go func() { strassenParallel(a22, s4, p4); wg.Done() }()
		go func() { strassenParallel(s5, s6, p5); wg.Done() }()
		go func() { strassenParallel(s7, s8, p6); wg.Done() }()
		go func() { strassenParallel(s9, s10, p7); wg.Done() }()

		wg.Wait()
	} else {
		// Sequential fallback
		strassenSequential(a11, s1, p1)
		strassenSequential(s2, b22, p2)
		strassenSequential(s3, b11, p3)
		strassenSequential(a22, s4, p4)
		strassenSequential(s5, s6, p5)
		strassenSequential(s7, s8, p6)
		strassenSequential(s9, s10, p7)
	}

	// Combine
	add(p5, p4, c11)
	sub(c11, p2, c11)
	add(c11, p6, c11)

	add(p1, p2, c12)
	add(p3, p4, c21)

	add(p5, p1, c22)
	sub(c22, p3, c22)
	sub(c22, p7, c22)
}

type Matrix [][]int

func strassen_old(A, B, C Matrix) {
	n := len(A)

	if n <= THRESHOLD {
		multiplyMatrixIterative(A, B, C)
		return
	}

	computeMatricesParallel_old(A, B, C, n)
}

func computeMatricesParallel_old(A, B, C Matrix, n int) {
	a11, a12, a21, a22 := splitView(A)
	b11, b12, b21, b22 := splitView(B)
	c11, c12, c21, c22 := splitView(C)

	half := n / 2

	// S matrices
	s := make([]Matrix, 10)
	for i := range s {
		s[i] = makeZeroMatrix(half)
	}

	// M matrices
	m := make([]Matrix, 7)
	for i := range m {
		m[i] = makeZeroMatrix(half)
	}

	// Compute S
	sub(b12, b22, s[0])
	add(a11, a12, s[1])
	add(a21, a22, s[2])
	sub(b21, b11, s[3])
	add(a11, a22, s[4])
	add(b11, b22, s[5])
	sub(a12, a22, s[6])
	add(b21, b22, s[7])
	sub(a11, a21, s[8])
	add(b11, b12, s[9])

	var wg sync.WaitGroup
	wg.Add(7)
	go func() { strassen_old(a11, s[0], m[0]); wg.Done() }()
	go func() { strassen_old(s[1], b22, m[1]); wg.Done() }()
	go func() { strassen_old(s[2], b11, m[2]); wg.Done() }()
	go func() { strassen_old(a22, s[3], m[3]); wg.Done() }()
	go func() { strassen_old(s[4], s[5], m[4]); wg.Done() }()
	go func() { strassen_old(s[6], s[7], m[5]); wg.Done() }()
	go func() { strassen_old(s[8], s[9], m[6]); wg.Done() }()
	wg.Wait()

	// Combine
	add(m[4], m[3], c11)
	sub(c11, m[1], c11)
	add(c11, m[5], c11)

	add(m[0], m[1], c12)

	add(m[2], m[3], c21)

	add(m[4], m[0], c22)
	sub(c22, m[2], c22)
	sub(c22, m[6], c22)
}

func strassen(A, B, C Matrix) {
	n := len(A)

	if n <= THRESHOLD {
		multiplyMatrixIterative(A, B, C)
	} else {
		matrices := computeMatricesParallel(A, B, n)
		combineResult(matrices, C)
	}
}

func computeMatricesParallel(A, B Matrix, n int) [7]Matrix {
	a11, a12, a21, a22 := splitView(A)
	b11, b12, b21, b22 := splitView(B)

	half := n / 2

	// S matrices
	S := [10]Matrix{}
	for i := range S {
		S[i] = makeZeroMatrix(half)
	}

	// Compute S
	sub(b12, b22, S[0])
	add(a11, a12, S[1])
	add(a21, a22, S[2])
	sub(b21, b11, S[3])
	add(a11, a22, S[4])
	add(b11, b22, S[5])
	sub(a12, a22, S[6])
	add(b21, b22, S[7])
	sub(a11, a21, S[8])
	add(b11, b12, S[9])

	// M matrices
	M := [7]Matrix{}
	for i := range M {
		M[i] = makeZeroMatrix(half)
	}

	var wg sync.WaitGroup
	wg.Add(7)
	go func() { strassen(a11, S[0], M[0]); wg.Done() }()
	go func() { strassen(S[1], b22, M[1]); wg.Done() }()
	go func() { strassen(S[2], b11, M[2]); wg.Done() }()
	go func() { strassen(a22, S[3], M[3]); wg.Done() }()
	go func() { strassen(S[4], S[5], M[4]); wg.Done() }()
	go func() { strassen(S[6], S[7], M[5]); wg.Done() }()
	go func() { strassen(S[8], S[9], M[6]); wg.Done() }()
	wg.Wait()

	return M
}

func combineResult(m [7]Matrix, C Matrix) {
	c11, c12, c21, c22 := splitView(C)

	add(m[4], m[3], c11)
	sub(c11, m[1], c11)
	add(c11, m[5], c11)

	add(m[0], m[1], c12)

	add(m[2], m[3], c21)

	add(m[4], m[0], c22)
	sub(c22, m[2], c22)
	sub(c22, m[6], c22)
}

// func mergeQuadrants(c11, c12, c21, c22 Matrix) Matrix {
// 	half := len(c11)
// 	n := half * 2

// 	C := make([][]int, n)
// 	for i := range C {
// 		C[i] = make([]int, n)
// 	}

// 	for i := range n {
// 		for j := range n {
// 			C[i][j] = c11[i][j]
// 			C[i][j+half] = c12[i][j]
// 			C[i+half][j] = c11[i][j]
// 			C[i+half][j+half] = c11[i][j]
// 		}
// 	}

// 	return C
// }

File: rust/iterative/src/matrix/strassen.rs
use crate::matrix::{
    Matrix, mat_mul::multiply_matrix, matrix::{MatrixView, THRESHOLD, add_vec, merge_quadrants, split_view}, matrix_op::sub_vec
};

pub fn strassen_sequential(a: MatrixView, b: MatrixView) -> Matrix {
    let n = a.rows;

    // Base case: small matrix, use standard multiplication
    if n <= THRESHOLD {
        return multiply_matrix(a, b); // your standard iterative matmul
    }

    let half = n / 2;

    // Split A and B into 4 submatrices each
    let a11 = a.sub_view(0, 0, half, half);
    let a12 = a.sub_view(0, half, half, half);
    let a21 = a.sub_view(half, 0, half, half);
    let a22 = a.sub_view(half, half, half, half);

    let b11 = b.sub_view(0, 0, half, half);
    let b12 = b.sub_view(0, half, half, half);
    let b21 = b.sub_view(half, 0, half, half);
    let b22 = b.sub_view(half, half, half, half);

    // Strassen's 7 products
    let m1 = strassen_sequential(
        add_vec(&a11.to_matrix(), &a22.to_matrix()).view(),
        add_vec(&b11.to_matrix(), &b22.to_matrix()).view(),
    );
    let m2 = strassen_sequential(add_vec(&a21.to_matrix(), &a22.to_matrix()).view(), b11);
    let m3 = strassen_sequential(a11, sub_vec(&b12, &b22).view());
    let m4 = strassen_sequential(a22, sub_vec(&b21, &b11).view());
    let m5 = strassen_sequential(add_vec(&a11.to_matrix(), &a12.to_matrix()).view(), b22);
    let m6 = strassen_sequential(
        sub_vec(&a21, &a11).view(),
        add_vec(&b11.to_matrix(), &b12.to_matrix()).view(),
    );
    let m7 = strassen_sequential(
        sub_vec(&a12, &a22).view(),
        add_vec(&b21.to_matrix(), &b22.to_matrix()).view(),
    );

    // Combine into result submatrices
    let c11 = add_vec(&sub_vec(&add_vec(&m1, &m4).view(), &m5.view()), &m7);
    let c12 = add_vec(&m3, &m5);
    let c21 = add_vec(&m2, &m4);
    let c22 = add_vec(&sub_vec(&add_vec(&m1, &m3).view(), &m2.view()), &m6);

    // Merge submatrices into final result
    let mut c = Matrix::new(n, n);

    for i in 0..half {
        for j in 0..half {
            c.set(i, j, c11.get(i, j));
            c.set(i, j + half, c12.get(i, j));
            c.set(i + half, j, c21.get(i, j));
            c.set(i + half, j + half, c22.get(i, j));
        }
    }

    c
}

pub fn strassen(a: MatrixView, b: MatrixView) -> Matrix {
    let n = a.rows;

    if n <= THRESHOLD {
        return multiply_matrix(a, b);
    }

    let a_blocks = split_view(a);
    let b_blocks = split_view(b);
    let matrices = compute_matrices_parallel(a_blocks, b_blocks);
    combine_result(matrices)
}

fn compute_matrices_parallel(
    a: (MatrixView, MatrixView, MatrixView, MatrixView),
    b: (MatrixView, MatrixView, MatrixView, MatrixView),
) -> [Matrix; 7] {
    let (a11, a12, a21, a22) = a;
    let (b11, b12, b21, b22) = b;

    let mut m1 = Matrix::new(a11.rows, a11.cols);
    let mut m2 = Matrix::new(a11.rows, a11.cols);
    let mut m3 = Matrix::new(a11.rows, a11.cols);
    let mut m4 = Matrix::new(a11.rows, a11.cols);
    let mut m5 = Matrix::new(a11.rows, a11.cols);
    let mut m6 = Matrix::new(a11.rows, a11.cols);
    let mut m7 = Matrix::new(a11.rows, a11.cols);

    rayon::scope(|s| {
        s.spawn(|_| m1 = strassen(
            add_vec(&a11.to_matrix(), &a22.to_matrix()).view(), 
            add_vec(&b11.to_matrix(), &b22.to_matrix()).view()
        ));
        s.spawn(|_| m2 = strassen(
            add_vec(&a21.to_matrix(), &a22.to_matrix()).view(), 
            b11
        ));
        s.spawn(|_| m3 = strassen(
            a11, 
            sub_vec(&b12, &b22).view()
        ));
        s.spawn(|_| m4 = strassen(
            a22, 
            sub_vec(&b21, &b11).view()
        ));
        s.spawn(|_| m5 = strassen(
            add_vec(&a11.to_matrix(), &a12.to_matrix()).view(), 
            b22
        ));
        s.spawn(|_| m6 = strassen(
            sub_vec(&a21, &a11).view(), 
            add_vec(&b11.to_matrix(), &b12.to_matrix()).view()
        ));
        s.spawn(|_| m7 = strassen(
            sub_vec(&a12, &a22).view(), 
            add_vec(&b21.to_matrix(), &b22.to_matrix()).view()
        ));
    });

    [m1, m2, m3, m4, m5, m6, m7]
}

fn combine_result(m: [Matrix; 7]) -> Matrix {
    let c11 = add_vec(&sub_vec(&add_vec(&m[0], &m[3]).view(), &m[4].view()), &m[6]);
    let c12 = add_vec(&m[2], &m[4]);
    let c21 = add_vec(&m[1], &m[3]);
    let c22 = add_vec(&sub_vec(&add_vec(&m[0], &m[2]).view(), &m[1].view()), &m[5]);

    merge_quadrants(&c11, &c12, &c21, &c22)
}